File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.py
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        N = 0
        dummy = ListNode(0, head)
        curr = head
        while curr:
            N += 1
            curr = curr.next

        # Remove N - n from the linked list
        cnt = 0
        curr = dummy
        while curr.next:
            if cnt == (N - n):
                curr.next = curr.next.next
                break
            else:
                curr = curr.next
            cnt += 1

        return dummy.next

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        int k = lists.length;
        if (k == 0) return null;
        
        for (int i = 1; i < k; i++) {
            lists[0] = mergeTwoLists(lists[0], lists[i]);
        }

        return lists[0];
    }

    public ListNode mergeTwoLists(ListNode node1, ListNode node2) {
        ListNode dummy = new ListNode(-1);
        ListNode current = dummy;

        while (node1 != null && node2 != null) {
            if (node1.val <= node2.val) {
                current.next = node1;
                node1 = node1.next;
            } else {
                current.next = node2;
                node2 = node2.next;
            }
            current = current.next;
        }
        current.next = (node1 != null) ? node1 : node2;
        return dummy.next;
    }
}
